ADBDEV-4726-33 Check that wfile is not null

dnskvlnk opened the pull request:
Check that wfile is not null

agg_hash_reload passes wfile to BufFileRead where it's dereferenced without
checking that it's not null. We can't guarantee that file is always valid.
This patch replaces redundant null-check with Assert

Commit:
Check that file is not null

Commit:
Check that file is not null

BufFileRead dereferences file without checking that it's not null.
This patch adds an assert

RekGRpth commented on src/backend/storage/file/buffile.c:
```diff
@@ -490,6 +490,8 @@ BufFileRead(BufFile *file, void *ptr, size_t size)
 	size_t		nread = 0;
 	size_t		nthistime;
 
+	Assert(file);
```
This assertion will not check anything, because it is already asserted in the readHashEntry function. In addition, the static analyzer does not see assertions and therefore will still issue the same warning.

dnskvlnk replied:
Since it's not obvious that `file` is correct I suggest using `Insist` or stick with `elog(ERROR` which I already did

RekGRpth replied:
Try use Insist, because for elog(ERROR you should provide test case.

dnskvlnk replied:
Done

Commit:
Substitute assert with runtime check

Commit:
Check that file is not null

BufFileRead dereferences file without checking that it's not null.
We can't guarantee that file is always valid. This patch adds Insist

RekGRpth commented on src/backend/storage/file/buffile.c:
```diff
@@ -490,6 +490,8 @@ BufFileRead(BufFile *file, void *ptr, size_t size)
 	size_t		nread = 0;
 	size_t		nthistime;
 
+	Insist(file);
```
I think you should replace condition https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/executor/execHHashagg.c#L1881 by assert

dnskvlnk replied:
Done

RekGRpth replied:
I think this insist is redundant

RekGRpth replied:
I still think so

dnskvlnk replied:
Done

Commit:
Check that file is not null

BufFileRead dereferences file without checking that it's not null.
We can't guarantee that file is always valid. This patch adds Insist
and replaces redundant null-check with Assert

RekGRpth commented on src/backend/executor/execHHashagg.c:
```diff
@@ -1878,13 +1878,10 @@ agg_hash_reload(AggState *aggstate)
-		spill_file->file_info->suspended = false;
-		hashtable->mem_for_metadata  += FREEABLE_BATCHFILE_METADATA;
-	}
+	Assert(spill_file->file_info->wfile && spill_file->file_info->suspended);
```
why not simply
```suggestion
Assert(spill_file->file_info->wfile);
Assert(spill_file->file_info->suspended);
```

dnskvlnk replied:
What's the difference?

RekGRpth replied:
gihub.dev shows less diff

dnskvlnk replied:
I don't see any sense splitting it into two different assertions since they refer to the same object

RekGRpth commented:
update PR description

dnskvlnk commented:
@RekGRpth it's up to date

Commit:
Check that wfile is not null

agg_hash_reload passes wfile to BufFileRead where it's dereferenced without
checking that it's not null. We can't guarantee that file is always valid.
This patch replaces redundant null-check with Assert
